src/unstract/sdk/file_storage/fs_permanent.py

```diff
@@ -32,24 +34,29 @@ def __init__(
             super().__init__(provider)
         else:
             raise NotImplementedError
-        self.legacy_storage_path = legacy_storage_path
 
-    def _copy_on_write(self, path):
-        """Copies the file to the lazily. Checks if the file is present in the
-        Local File system. If yes, copies the file to the mentioned path using
-        the remote file system. This is a silent copy done on need basis.
+    def _copy_on_read(self, path: str, legacy_storage_path: str):
+        """Copies the file to the remote storage lazily if not present already.
+        Checks if the file is present in the Local File system. If yes, copies
+        the file to the mentioned path using the remote file system. This is a
+        silent copy done on need basis.
 
         Args:
             path (str): Path to the file
+            legacy_storage_path (str): Legacy path to the same file
 
         Returns:
             NA
         """
+        # If path does not exist on remote storage
         if not self.exists(path):
             local_file_storage = FileStorage(provider=FileStorageProvider.LOCAL)
-            local_file_path = self.legacy_storage_path + "/" + path
-            if local_file_storage.exists(local_file_path):
-                self.upload(local_file_path, path)
+            if legacy_storage_path:
```

Comment:
```
local_file_storage = FileStorage(provider=FileStorageProvider.LOCAL)
if not self.exists(path) and local_file_storage.exists(local_file_path):
    self.upload(local_file_path, path)

```

@gaya3-zipstack Implying [this](https://github.com/Zipstack/unstract-sdk/pull/118/files#r1851860662) changes, the suggestion might be more readable, feel free to ignore if you feel otherwise.

Reply:
Done/